components/scheduler/scheduler.go

```diff
@@ -0,0 +1,494 @@
+	}
+}
+
+func searchCache(cache *schedspb.NextTimeCache, after time.Time) (scheduler.GetNextTimeResult, bool) {
+	// The cache covers a contiguous time range so we can do a linear search in it.
+	start := cache.StartTime.AsTime()
+	afterOffset := int64(after.Sub(start))
+	for i, nextOffset := range cache.NextTimes {
+		if nextOffset > afterOffset {
+			next := start.Add(time.Duration(nextOffset))
+			nominal := next
+			if i < len(cache.NominalTimes) && cache.NominalTimes[i] != 0 {
+				nominal = start.Add(time.Duration(cache.NominalTimes[i]))
+			}
+			return scheduler.GetNextTimeResult{Nominal: nominal, Next: next}, true
+		}
+	}
+	// Ran off end: if completed, then we're done
+	if cache.Completed {
+		return scheduler.GetNextTimeResult{}, true
+	}
+	return scheduler.GetNextTimeResult{}, false
+}
+
+func (s *Scheduler) getNextTime(after time.Time) scheduler.GetNextTimeResult {
+	// Asking for a time before the cache, need to refill.
+	// Also if version changed (so we can fix a bug immediately).
+	if s.nextTimeCache == nil ||
+		after.Before(s.nextTimeCache.StartTime.AsTime()) {
+		s.fillNextTimeCache(after)
+	}
+
+	// We may need up to three tries: the first is in the cache as it exists now,
+	// the second is refilled from cacheBase, and the third is if cacheBase was set
+	// too far in the past, we ignore it and fill the cache from after.
+	for try := 1; try <= 3; try++ {
+		if res, ok := searchCache(s.nextTimeCache, after); ok {
+			return res
+		}
+		// Otherwise refill from base
+		s.fillNextTimeCache(after)
+	}
+
+	// This should never happen unless there's a bug.
+	s.logger.Error("getNextTime: time not found in cache", tag.NewTimeTag("after", after))
+	return scheduler.GetNextTimeResult{}
+}
+
+func (s *Scheduler) bufferWorkflowStart(nominalTime, actualTime time.Time, overlapPolicy enumspb.ScheduleOverlapPolicy, manual bool) {
+	s.logger.Debug("bufferWorkflowStart", tag.NewTimeTag("start-time", nominalTime), tag.NewTimeTag("actual-start-time", actualTime),
+		tag.NewAnyTag("overlap-policy", overlapPolicy), tag.NewBoolTag("manual", manual))
+	if s.tweakables.MaxBufferSize > 0 && len(s.Args.State.BufferedStarts) >= s.tweakables.MaxBufferSize {
+		s.logger.Warn("Buffer too large", tag.NewTimeTag("start-time", nominalTime), tag.NewAnyTag("overlap-policy", overlapPolicy), tag.NewBoolTag("manual", manual))
+		s.metrics.Counter(metrics.ScheduleBufferOverruns.Name()).Record(1)
+		s.Args.Info.BufferDropped += 1
+		return
+	}
+	s.Args.State.BufferedStarts = append(s.Args.State.BufferedStarts, &schedspb.BufferedStart{
+		NominalTime:   timestamppb.New(nominalTime),
+		ActualTime:    timestamppb.New(actualTime),
+		OverlapPolicy: overlapPolicy,
+		Manual:        manual,
+	})
+}
+
+func (s *Scheduler) processTimeRange(
+	start, end time.Time,
+	overlapPolicy enumspb.ScheduleOverlapPolicy,
+	manual bool,
+	limit *int,
+) (time.Time, time.Time) {
+	s.logger.Debug("processTimeRange", tag.NewTimeTag("start", start), tag.NewTimeTag("end", end),
```

Comment:
what about paused? this isn't checking canTakeScheduledAction at all?

we should check it in the loop and also check it before the loop to skip over everything on unpause after a very long interval

Reply:
Will be added in the next PR --- to properly pause/unpause schedules we need to be able to signal state machines which is not yet implemented in this PR.